src/linear_algebra/unitary_group.lean

```diff
@@ -0,0 +1,176 @@
+{ F   := λ _, n → n → α,
+  coe := λ A, A.val }
+
+/--
+  `to_lin' A` is matrix multiplication of vectors by `A`, as a linear map.
+
+  After the group structure on `unitary_group n` is defined,
+  we show in `to_linear_equiv` that this gives a linear equivalence.
+-/
+def to_lin' (A : unitary_group n α) := matrix.to_lin' A
+
+lemma ext_iff (A B : unitary_group n α) : A = B ↔ ∀ i j, A i j = B i j :=
+iff.trans subtype.ext_iff_val ⟨(λ h i j, congr_fun (congr_fun h i) j), matrix.ext⟩
+
+@[ext] lemma ext (A B : unitary_group n α) : (∀ i j, A i j = B i j) → A = B :=
+(unitary_group.ext_iff A B).mpr
+
+instance : has_inv (unitary_group n α) :=
+⟨λ A, ⟨star A.1, matrix.nonsing_inv_left_right _ _ $
+  by { dsimp only, rw [star_star, A.2]}⟩⟩
+
+instance : has_mul (unitary_group n α) :=
+⟨λ A B, ⟨A.1 ⬝ B.1, by rw [star_mul, ←matrix.mul_assoc,
+                                     matrix.mul_assoc _ _ A.1, A.2, matrix.mul_one, B.2],⟩⟩
+
+instance : has_one (unitary_group n α) :=
+⟨⟨1, by rw [star_one, matrix.one_mul]⟩⟩
+
+instance : inhabited (unitary_group n α) := ⟨1⟩
+
+section coe_lemmas
+
+variables (A B : unitary_group n α)
+
+@[simp] lemma inv_val : ↑(A⁻¹) = (star A : matrix n n α) := rfl
+
+@[simp] lemma inv_apply : ⇑(A⁻¹) = (star A : matrix n n α) := rfl
+
+@[simp] lemma mul_val : ↑(A * B) = A ⬝ B := rfl
+
+@[simp] lemma mul_apply : ⇑(A * B) = (A ⬝ B) := rfl
+
+@[simp] lemma one_val : ↑(1 : unitary_group n α) = (1 : matrix n n α) := rfl
+
+@[simp] lemma one_apply : ⇑(1 : unitary_group n α) = (1 : matrix n n α) := rfl
+
+@[simp] lemma to_lin'_mul :
+  to_lin' (A * B) = (to_lin' A).comp (to_lin' B) :=
+matrix.to_lin'_mul A B
+
+@[simp] lemma to_lin'_one :
+  to_lin' (1 : unitary_group n α) = linear_map.id :=
+matrix.to_lin'_one
+
+end coe_lemmas
+
+instance : group (unitary_group n α) :=
+{ mul_assoc := λ A B C, subtype.eq (A.val.mul_assoc B.val C.val),
+  one_mul := λ M, subtype.eq $ one_mul _,
+  mul_one := λ M, subtype.eq $ mul_one _,
+  mul_left_inv := λ A, subtype.eq A.2,
+  ..unitary_group.has_mul,
+  ..unitary_group.has_inv,
+  ..unitary_group.has_one }
+
+/-- `to_linear_equiv A` is matrix multiplication of vectors by `A`, as a linear equivalence. -/
+def to_linear_equiv (A : unitary_group n α) : (n → α) ≃ₗ[α] (n → α) :=
+{ inv_fun := A⁻¹.to_lin',
+  left_inv := λ x, calc
+    A⁻¹.to_lin'.comp A.to_lin' x
+        = (A⁻¹ * A).to_lin' x : by rw [←to_lin'_mul]
+    ... = x : by rw [mul_left_inv, to_lin'_one, id_apply],
+  right_inv := λ x, calc
+    A.to_lin'.comp A⁻¹.to_lin' x
+        = (A * A⁻¹).to_lin' x : by rw [←to_lin'_mul]
+    ... = x : by rw [mul_right_inv, to_lin'_one, id_apply],
+  ..matrix.to_lin' A }
+
+/-- `to_GL` is the map from the unitary group to the general linear group -/
+def to_GL (A : unitary_group n α) : general_linear_group α (n → α) :=
+general_linear_group.of_linear_equiv (to_linear_equiv A)
+
+lemma coe_to_GL (A : unitary_group n α) :
+  (@coe (units _) _ _ (to_GL A)) = A.to_lin' :=
```

Comment:
Can you write this as something more readable like the following?
```suggestion
  (to_GL A : units _) = A.to_lin' :=
```

Reply:
Nope, that doesn't seem to work.

Reply:
This does
```suggestion
  ↑(to_GL A) = A.to_lin' :=
```

Reply:
This change looks good.

Reply:
@shingtaklam1324 this comment still applies

Reply:
done.